### ISSUE TITLE
Data Race: concurrent access of error state during duplexHTTPCall

### ISSUE THREAD

**Issue by rodaine** [connectrpc/otelconnect-go#171]
I'm having difficulty determining the provenance of this particular race, but adjacent to the `closeSpan` call alongside a send func (similar to #170):

```
WARNING: DATA RACE
Read at 0x00c00fc5b408 by goroutine 34631:
  connectrpc.com/connect.(*Error).Unwrap()
      /Users/rodaine/pkg/mod/connectrpc.com/connect@v1.16.1/error.go:210 +0x38
  errors.is()
      /Users/rodaine/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.3.darwin-arm64/src/errors/wrap.go:63 +0x24c
  errors.Is()
      /Users/rodaine/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.3.darwin-arm64/src/errors/wrap.go:50 +0xfc
  connectrpc.com/connect.IsNotModifiedError()
      /Users/rodaine/pkg/mod/connectrpc.com/connect@v1.16.1/error.go:264 +0x68
  connectrpc.com/otelconnect.statusCodeAttribute()
      /Users/rodaine/pkg/mod/connectrpc.com/otelconnect@v0.7.0/attributes.go:110 +0x5e4
  connectrpc.com/otelconnect.(*Interceptor).WrapStreamingClient.func1.2()
      /Users/rodaine/pkg/mod/connectrpc.com/otelconnect@v0.7.0/interceptor.go:234 +0x120

Previous write at 0x00c00fc5b408 by goroutine 34378:
  connectrpc.com/connect.NewError()
      /Users/rodaine/pkg/mod/connectrpc.com/connect@v1.16.1/error.go:133 +0xcc
  connectrpc.com/connect.wrapIfContextError()
      /Users/rodaine/pkg/mod/connectrpc.com/connect@v1.16.1/error.go:306 +0x114
  connectrpc.com/connect.(*duplexHTTPCall).makeRequest()
      /Users/rodaine/pkg/mod/connectrpc.com/connect@v1.16.1/duplex_http_call.go:310 +0x2f8
  connectrpc.com/connect.(*duplexHTTPCall).Send.gowrap1()
      /Users/rodaine/pkg/mod/connectrpc.com/connect@v1.16.1/duplex_http_call.go:108 +0x38

Goroutine 34631 (running) created at:
  context.(*afterFuncCtx).cancel.func1()
      /Users/rodaine/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.3.darwin-arm64/src/context/context.go:348 +0x44
  sync.(*Once).doSlow()
      /Users/rodaine/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.3.darwin-arm64/src/sync/once.go:74 +0x150
  sync.(*Once).Do()
      /Users/rodaine/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.3.darwin-arm64/src/sync/once.go:65 +0x58
  context.(*afterFuncCtx).cancel()
      /Users/rodaine/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.3.darwin-arm64/src/context/context.go:347 +0x174
  context.(*cancelCtx).cancel()
      /Users/rodaine/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.3.darwin-arm64/src/context/context.go:558 +0x348
  context.(*cancelCtx).cancel()
      /Users/rodaine/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.3.darwin-arm64/src/context/context.go:558 +0x348
  context.(*cancelCtx).cancel()
      /Users/rodaine/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.3.darwin-arm64/src/context/context.go:558 +0x348
  context.WithCancel.func1()
      /Users/rodaine/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.3.darwin-arm64/src/context/context.go:237 +0x78
  main.run.func1()
      main.go:53 +0x20c

Goroutine 34378 (finished) created at:
  connectrpc.com/connect.(*duplexHTTPCall).Send()
      /Users/rodaine/pkg/mod/connectrpc.com/connect@v1.16.1/duplex_http_call.go:108 +0x33c
  connectrpc.com/connect.(*envelopeWriter).write()
      /Users/rodaine/pkg/mod/connectrpc.com/connect@v1.16.1/envelope.go:212 +0x88
  connectrpc.com/connect.(*envelopeWriter).Write()
      /Users/rodaine/pkg/mod/connectrpc.com/connect@v1.16.1/envelope.go:157 +0x37c
  connectrpc.com/connect.(*envelopeWriter).marshalAppend()
      /Users/rodaine/pkg/mod/connectrpc.com/connect@v1.16.1/envelope.go:195 +0x564
  connectrpc.com/connect.(*envelopeWriter).Marshal()
      /Users/rodaine/pkg/mod/connectrpc.com/connect@v1.16.1/envelope.go:143 +0x114
  connectrpc.com/connect.(*connectStreamingClientConn).Send()
      /Users/rodaine/pkg/mod/connectrpc.com/connect@v1.16.1/protocol_connect.go:589 +0x54
  connectrpc.com/connect.(*errorTranslatingClientConn).Send()
      /Users/rodaine/pkg/mod/connectrpc.com/connect@v1.16.1/protocol.go:206 +0x7c
  connectrpc.com/otelconnect.(*streamingState).send()
      /Users/rodaine/pkg/mod/connectrpc.com/otelconnect@v0.7.0/streaming.go:102 +0x68
  connectrpc.com/otelconnect.(*Interceptor).WrapStreamingClient.func1.5()
      /Users/rodaine/pkg/mod/connectrpc.com/otelconnect@v0.7.0/interceptor.go:262 +0x10c
  connectrpc.com/otelconnect.(*streamingClientInterceptor).Send()
      /Users/rodaine/pkg/mod/connectrpc.com/otelconnect@v0.7.0/payloadinterceptor.go:34 +0xa8
  connectrpc.com/connect.(*ClientStreamForClient[ ... ]).Send()
      /Users/rodaine/pkg/mod/connectrpc.com/connect@v1.16.1/client_stream.go:69 +0x110
```

**Comment by emcfarlane** [comment]
I think this is the same race as #170 . Client stream on close accesses the state of the error here: https://github.com/connectrpc/otelconnect-go/blob/0518f328245e8a6dd42c08eee6cb3e90fbca3cea/interceptor.go#L234
Which is caused from a context cancellation that races with a receive of an error.

**Comment by jhump** [comment]
Resolved in #173, included in release v0.7.1.
